designs/2018-relative-package-loading/README.md

```diff
@@ -0,0 +1,278 @@
+
+I consider this to be acceptable because end users are already exposed to many of the details of their shareable configs, in that changes to shareable configs will lead to different linting errors/warnings being reported on the end user's code. From an encapsulation perspective, it seems important to protect configs from caring about details of *sibling* shareable configs, since this would create implicit dependencies between the sibling configs where no dependency otherwise existed. However, it seems less important to protect configs from details of *descendent* shareable configs, since there is already a dependency relationship between them anyway.
+
+Additionally, the problem of exposure to details of dependency config appears to be inherent to any solution that both (a) gives users the power to arbitrarily override third-party configuration, and (b) allows third-party configuration to pull in custom rules from multiple external sources.
+
+### Hierarchial rule name resolution adds complexity
+
+Hierarchial rule name resolution would increase the complexity of how configs are resolved, potentially adding a maintenance burden. (However, the implementation would also provide a good opportunity to refactor the existing config-loading code, which might make that code easier to maintain overall.)
+
+## Backwards Compatibility Analysis
+
+This proposal maintains compatibility for most shareable configs, and most local installation setups from end users. There are a few backwards-incompatible parts:
+
+* Most "global installation" setups that use plugins will need to be modified. Previously, a user need to install plugins globally when ESLint was installed globally. With this proposal implemented, a user should install plugins as a dependency of the codebase that contains their config file.
+* Configs can no longer rely on plugin names being globally unique. For example, if two configs independently configure a plugin with the same name, their configurations would previously override each other; with this proposal implemented, the configs will create two independent configurations. Along similar lines, shareable configs can no longer reconfigure their siblings' plugins, as described in the "Drawbacks" section.
+* Previously, a config could use something like `extends: ['plugin:foo/bar']` without explicitly declaring a dependency on the `eslint-plugin-foo` plugin. With this change, such a config would be required to also add `plugins: ['foo']` for ESLint to load `eslint-plugin-foo`.
```

Comment:
I think that this will be a tough breaking change. I know that most plugins have `plugins: ['foo']` in their recommended presets in order to allow omitting `plugins` property.

- [eslint-plugin-import](https://github.com/benmosher/eslint-plugin-import/blob/798eed7e559adab2eac07bf1b3e3518d4b7a4296/config/recommended.js#L6)
- [eslint-plugin-node](https://github.com/mysticatea/eslint-plugin-node/blob/932836b5a97e33c35ada5af0d3dd2d5ba1e2307b/lib/configs/recommended.json#L13)
- [eslint-plugin-react](https://github.com/yannickcr/eslint-plugin-react/blob/a33bfb18ca515e824c5ade55aab1505877b4501e/index.js#L117)
- [eslint-plugin-jsx-a11y](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/e53906d0b2a9402b019625349ed2d58d178b3239/src/index.js#L42)
- [eslint-plugin-vue](https://github.com/vuejs/eslint-plugin-vue/blob/9fdf8e030c880e99f9bfa63fe10c3408961ee256/lib/configs/base.js#L19)
- [eslint-plugin-jest](https://github.com/jest-community/eslint-plugin-jest/blob/73c9187be44fa73077775c0c85a136d4d65aa214/index.js#L36)
- [eslint-plugin-eslint-comments](https://github.com/mysticatea/eslint-plugin-eslint-comments/blob/c427c1c0577fab56f2ff0919053d249938507c83/lib/configs/recommended.js#L5)
- [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier/blob/8675d5713f5171981119b89c2e8a58fda6b81259/eslint-plugin-prettier.js#L111)

Reply:
Hmm, I hadn't realized this was so common.

I suppose we could make it so that `extends: ['plugin:foo/bar']` automatically implies `plugins: ['foo']` in the config where it's used. I was hoping to simplify this aspect of plugin-loading but it doesn't seem doable at the moment.

Reply:
I've updated the proposal to implement https://github.com/eslint/rfcs/pull/5#discussion_r238535687, so this breaking change is no longer part of the proposal.